Avoid potential NullPointerException

mathieu-pousse opened the pull request:
Hi there, 

For an unknown reason (so far), I am in a case where the buffer is null.

It result in an NPE.

```
2024-03-07 14:19:31,682 ERROR [org.infinispan.CLUSTER] (jgroups-14,f4d5439b3910-35958) ISPN000474: Error processing request 0@918a36bef812-46695: java.lang.NullPointerException: Cannot read the array length because "this.bytes" is null
	at org.infinispan.marshall.core.AbstractBytesObjectInput.checkPosLength(AbstractBytesObjectInput.java:260)
	at org.infinispan.marshall.core.AbstractBytesObjectInput.readByte(AbstractBytesObjectInput.java:107)
	at org.infinispan.marshall.core.AbstractBytesObjectInput.readUnsignedByte(AbstractBytesObjectInput.java:113)
	at org.infinispan.marshall.core.GlobalMarshaller.readNullableObject(GlobalMarshaller.java:356)
	at org.infinispan.marshall.core.GlobalMarshaller.objectFromObjectInput(GlobalMarshaller.java:191)
	at org.infinispan.marshall.core.GlobalMarshaller.objectFromByteBuffer(GlobalMarshaller.java:220)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processRequest(JGroupsTransport.java:1536)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1476)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.lambda$up$1(JGroupsTransport.java:1695)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1687)
	at org.jgroups.JChannel.up(JChannel.java:749)
	at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:939)
	at org.jgroups.protocols.FRAG3.up(FRAG3.java:154)
	at org.jgroups.protocols.VERIFY_SUSPECT.up(VERIFY_SUSPECT.java:146)
	at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:193)
	at org.jgroups.protocols.FRAG2.up(FRAG2.java:161)
	at org.jgroups.protocols.FlowControl.up(FlowControl.java:319)
	at org.jgroups.protocols.FlowControl.up(FlowControl.java:319)
	at org.jgroups.protocols.pbcast.GMS.up(GMS.java:859)
	at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:246)
	at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:498)
	at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:722)
	at org.jgroups.protocols.VERIFY_SUSPECT2.up(VERIFY_SUSPECT2.java:119)
	at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:193)
	at org.jgroups.stack.Protocol.up(Protocol.java:372)
	at org.jgroups.protocols.MERGE3.up(MERGE3.java:288)
	at org.jgroups.protocols.Discovery.up(Discovery.java:314)
	at org.jgroups.protocols.RED.up(RED.java:119)
	at org.jgroups.protocols.TP.passBatchUp(TP.java:1210)
	at org.jgroups.util.MaxOneThreadPerSender$BatchHandlerLoop.passBatchUp(MaxOneThreadPerSender.java:287)
	at org.jgroups.util.SubmitToThreadPool$BatchHandler.run(SubmitToThreadPool.java:157)
	at org.jgroups.util.MaxOneThreadPerSender$BatchHandlerLoop.run(MaxOneThreadPerSender.java:276)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

I propose this change to properly handle this case instead of an unhandled null pointer exception

I bet the issue comes from the underlying jgroup, but this will be a safer way to handle those data.

Let me know

pruivo commented:
NullPointerException is not expected since Infinispan never sends null byte arrays. I would rather fix the bug than ignore the NPE. Do you have a reproducer?

mathieu-pousse commented:
I see.

I don't know how to reproduce, I mean, to give your a reproducible path.

We have a cluster of Keycloak that ships Infinispan (14.0.21).

We have stopped one node, and when we try to restart the node, during the cluster re balancing, I get this error.

mathieu-pousse commented:
Here are the states when the exception is triggered 

Last class from JGroup
![image](https://github.com/infinispan/infinispan/assets/2642023/2dd7f9cf-bb97-4163-82a5-faef37366f5f)

First class from infinispan marshaller
![image](https://github.com/infinispan/infinispan/assets/2642023/01d4299b-fb04-4490-aa8d-277ba81c78b3)

pruivo commented:
FD_SOCK2 messages should never reach Infinispan. I've noticed it is not the default JGroups stack. Can you share your JGroups configuration? @mathieu-pousse

mathieu-pousse commented:
Hum, interesting, here it is: 

```
    <jgroups>
        <!-- Allows custom jgroups configuration -->
        <stack name="tcp-ping" extends="tcp">
            <TCP external_addr="${env.JGROUPS_DISCOVERY_EXTERNAL_IP:127.0.0.1}"
                 bind_addr="${env.JGROUPS_TCP_BIND_ADDR:SITE_LOCAL}"
                 bind_port="7800"
                 recv_buf_size="${tcp.recv_buf_size:130k}"
                 send_buf_size="${tcp.send_buf_size:130k}"
                 sock_conn_timeout="300"
                 peer_addr_read_timeout="20000"
                 thread_pool.min_threads="0"
                 thread_pool.max_threads="100"
                 thread_pool.keep_alive_time="30000"/>
            <TCPPING initial_hosts="${env.JGROUPS_INITIAL_HOSTS}"
                     stack.combine="REPLACE" stack.position="MPING"/>
        </stack>
        <stack name="jdbc-ping" extends="tcp">
            <TCP external_addr="${env.JGROUPS_DISCOVERY_EXTERNAL_IP:127.0.0.1}"
                 bind_addr="${env.JGROUPS_TCP_BIND_ADDR:SITE_LOCAL}"
                 bind_port="7800"
                 recv_buf_size="130k"
                 send_buf_size="130k"
                 sock_conn_timeout="300"
                 peer_addr_read_timeout="20000"
                 thread_pool.min_threads="0"
                 thread_pool.max_threads="100"
                 thread_pool.keep_alive_time="30000"/>
            <JDBC_PING connection_driver="org.postgresql.Driver"
                       connection_username="${env.KC_DB_USERNAME}" connection_password="${env.KC_DB_PASSWORD}"
                       connection_url="${env.KC_DB_URL}"
                       initialize_sql="CREATE TABLE IF NOT EXISTS JGROUPSPING (own_addr varchar(200) NOT NULL, cluster_name varchar(200) NOT NULL, bind_addr varchar(200) NOT NULL, updated timestamp default current_timestamp, ping_data BYTEA, constraint PK_JGROUPSPING PRIMARY KEY (own_addr, cluster_name));"
                       insert_single_sql="INSERT INTO JGROUPSPING (own_addr, cluster_name, bind_addr, updated, ping_data) values (?, ?, '${env.JGROUPS_DISCOVERY_EXTERNAL_IP:127.0.0.1}', NOW(), ?);"
                       delete_single_sql="DELETE FROM JGROUPSPING WHERE own_addr=? AND cluster_name=?;"
                       select_all_pingdata_sql="SELECT ping_data, own_addr, cluster_name FROM JGROUPSPING WHERE cluster_name=?"
                       info_writer_sleep_time="500"
                       remove_all_data_on_view_change="true"
                       stack.combine="REPLACE"
                       stack.position="MPING" />
            <MERGE3 max_interval="30000"
                    min_interval="10000"/>
            <!-- Workaround for problems with FD_SOCK2, which might fail when the other nodes are not ready yet.
            See https://github.com/keycloak/keycloak/issues/21092 -->
            <FD_SOCK external_addr="${env.JGROUPS_DISCOVERY_EXTERNAL_IP}" stack.combine="REPLACE" stack.position="FD_SOCK2"/>
            <!-- Suspect node `timeout` to `timeout + timeout_check_interval` millis after the last heartbeat -->
            <FD_ALL timeout="10000"
                    interval="2000"
                    timeout_check_interval="1000"
            />
            <VERIFY_SUSPECT timeout="1500"  />
            <pbcast.NAKACK2 xmit_interval="500"
                            xmit_table_num_rows="100"
                            xmit_table_msgs_per_row="2000"
                            xmit_table_max_compaction_time="30000"
                            use_mcast_xmit="false"
                            discard_delivered_msgs="true"/>
            <UNICAST3 xmit_interval="500"
                      xmit_table_num_rows="100"
                      xmit_table_msgs_per_row="2000"
                      xmit_table_max_compaction_time="60000"
                      conn_expiry_timeout="0"/>
            <pbcast.STABLE desired_avg_gossip="50000"
                           max_bytes="4M"/>
            <pbcast.GMS print_local_addr="true" join_timeout="2000" print_view_details="true"/>
            <UFC max_credits="10M"
                 min_threshold="0.4"/>
            <MFC max_credits="10M"
                 min_threshold="0.4"/>
            <FRAG3/>
        </stack>
    </jgroups>
```

pruivo commented:
You have Keycloak nodes with different stacks in the same cluster. Your last screenshot shows an `EmptyMessage` with `FD_SOCK2` header but not `FD_SOCK2` in the thread stack. 

The Keycloak node with `tcp-ping` stack is sending `FD_SOCK2` related messages to another Keycloak node with `jdbc-ping`; since `jdbc-ping` does not use `FD_SOCK2`, the message is reaching Infinisan and it shouldn't. 

You should fix that and use the same stack in all the Keycloak nodes in the same cluster.

mathieu-pousse commented:
My bad, we use the jdbc-ping stack, the tcp-ping is not used (I copy pasted it but it is not used)

pruivo commented:
I would double-check your cluster or if you sent the correct screenshot because it is what your screenshot shows.

mathieu-pousse commented:
OK, my bad, got it :ok_hand: 
You are right, there is a misconfiguration on one node that messed up everything.

Thanks for opening my eyes, I'm sure one day I'll completely understand jgroups/* :sweat_smile: 

Thanks for the support, I appreciated that :+1:

pruivo commented:
Glad you figured it out.

>  I'm sure one day I'll completely understand jgroups

Let me know when you achieve this... I haven't yet. :rofl:
